src/main/java/com/walmartlabs/x12/standard/AbstractStandardX12Parser.java

```diff
@@ -0,0 +1,308 @@
+            if (segments.hasMoreSegmentLines()) {
+//                currentSegment = this.nextSegment(segmentIdx, segmentLines);
+                currentSegment = segments.peekSegment();
+                if (ISA_TRAILER_ID.equals(currentSegment.getSegmentIdentifier())) {
+                    this.parseInterchangeControlTrailer(currentSegment, x12Doc);
+                    currentSegment = segments.nextSegment();
+//                    segmentIdx++;
+                }
+            }
+        }
+    }
+
+    /**
+     * create a Standard X12 Document
+     * concrete parsers may override this as necessary to
+     * create a subclass of the StandardX12Document for
+     * use within the parser
+     * @return
+     */
+    protected StandardX12Document createX12Document() {
+        return new StandardX12Document();
+    }
+
+    protected abstract void parseTransasctionSet(List<X12Segment> transactionSegments, X12Group x12Group);
+
+    /**
+     * parse the ISA segment
+     *
+     * @param segment
+     * @param x12Doc
+     */
+    protected void parseInterchangeControlHeader(X12Segment segment, StandardX12Document x12Doc) {
+        LOGGER.debug(segment.getSegmentIdentifier());
+
+        String segmentIdentifier = segment.getSegmentIdentifier();
+        if (ISA_HEADER_ID.equals(segmentIdentifier)) {
+            InterchangeControlHeader isa = new InterchangeControlHeader();
+            isa.setAuthorizationInformationQualifier(segment.getSegmentElement(1));
+            isa.setAuthorizationInformation(segment.getSegmentElement(2));
+            isa.setSecurityInformationQualifier(segment.getSegmentElement(3));
+            isa.setSecurityInformation(segment.getSegmentElement(4));
+            isa.setInterchangeIdQualifier(segment.getSegmentElement(5));
+            isa.setInterchangeSenderId(segment.getSegmentElement(6));
+            isa.setInterchangeIdQualifier_2(segment.getSegmentElement(7));
+            isa.setInterchangeReceiverId(segment.getSegmentElement(8));
+            isa.setInterchangeDate(segment.getSegmentElement(9));
+            isa.setInterchangeTime(segment.getSegmentElement(10));
+            isa.setInterchangeControlStandardId(segment.getSegmentElement(11));
+            isa.setInterchangeControlVersion(segment.getSegmentElement(12));
+            isa.setInterchangeControlNumber(segment.getSegmentElement(13));
+            isa.setAcknowledgementRequested(segment.getSegmentElement(14));
+            isa.setUsageIndicator(segment.getSegmentElement(15));
+            isa.setElementSeparator(segment.getSegmentElement(16));
+
+            x12Doc.setInterchangeControlHeader(isa);
+        } else {
+            handleUnexpectedSegment(ISA_HEADER_ID, segmentIdentifier);
+        }
+    }
+
+    /**
+     * parse the ISE segment
+     *
+     * @param segment
+     * @param x12Doc
+     */
+    protected void parseInterchangeControlTrailer(X12Segment segment, StandardX12Document x12Doc) {
```

Comment:
Should we keep the return type consistent like `X12GroupHeader parseGroupHeader(X12Segment segment), InterchangeControlHeader parseInterchangeControlHeader(X12Segment segment)` etc.?
Meaning given X12Segment parse the respective segment and return it.
And then `standardParsingTemplate` would set respective retuned value.

Reply:
The abstract method that this class delegates to for transaction sets will require the subclass to add the objects to the group 

However, we need to keep track of the group we are on as we move thru the document